File: bindings/java/src/main/java/jp/golia/kevy/Subscriber.java
// Subscriber — the pub/sub consumer side (client-contract §3.11). A subscribed
// connection cannot send normal commands, so this is a distinct type;
// publishing is done from a normal KevyClient via publish(). Backends by URL:
// kevy:// / redis:// / tcp:// use a dedicated TCP socket; mem://<name> /
// file:// use the in-process bus. Anonymous mem:// is rejected (no other
// producer — use a named bus). recv() accepts both RESP2 arrays and RESP3
// push frames transparently.
package jp.golia.kevy;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

public final class Subscriber implements AutoCloseable {
    private RespConn remote;
    private EmbSub emb;
    private Duration readTimeout;

    private Subscriber(RespConn remote, EmbSub emb) {
        this.remote = remote;
        this.emb = emb;
    }

    /** Open a subscriber without subscribing yet. Anonymous mem:// is rejected. */
    public static Subscriber connect(String url) {
        Url.Target t = Url.parse(url);
        if (t instanceof Url.MemAnon) {
            throw new UnsupportedException("anonymous mem:// has no other producer; use mem://<name> for a shared bus");
        }
        if (t instanceof Url.Remote r) {
            RespConn conn = RespConn.dial(r.host(), r.port());
            return new Subscriber(conn, null);
        }
        return new Subscriber(null, new EmbSub(Registry.acquire(t)));
    }

    /** Open and subscribe in one step (≥1 channel). */
    public static Subscriber connectChannels(String url, byte[]... channels) {
        if (channels.length == 0) throw new InvalidInputException("connectChannels needs ≥ 1 channel");
        Subscriber s = connect(url);
        try {
            s.subscribe(channels);
        } catch (RuntimeException e) {
            s.close();
            throw e;
        }
        return s;
    }

    public void subscribe(byte[]... channels) {
        if (channels.length == 0) throw new InvalidInputException("SUBSCRIBE needs ≥ 1 channel");
        if (remote != null) remote.writeOnly(Argv.cmd("SUBSCRIBE").addAll(channels).list());
        else emb.add(channels, false);
    }

    public void psubscribe(byte[]... patterns) {
        if (patterns.length == 0) throw new InvalidInputException("PSUBSCRIBE needs ≥ 1 pattern");
        if (remote != null) remote.writeOnly(Argv.cmd("PSUBSCRIBE").addAll(patterns).list());
        else emb.add(patterns, true);
    }

    /** Unsubscribe from channels (empty = all). */
    public void unsubscribe(byte[]... channels) {
        if (remote != null) remote.writeOnly(Argv.cmd("UNSUBSCRIBE").addAll(channels).list());
        else emb.remove(channels, false);
    }

    public void punsubscribe(byte[]... patterns) {
        if (remote != null) remote.writeOnly(Argv.cmd("PUNSUBSCRIBE").addAll(patterns).list());
        else emb.remove(patterns, true);
    }

    /** Negotiate RESP3 push frames; remote-only, must precede any subscribe. */
    public PubsubEvent hello3() {
        if (remote == null) throw new UnsupportedException("HELLO 3 is remote-only; the embedded bus has no proto switch");
        Reply r = remote.request(Argv.cmd("HELLO").add("3").list());
        if (r.isError()) throw Errors.fromReplyText(r.payload());
        if (r instanceof Reply.Map || r.items() != null) {
            return new PubsubEvent(PubsubEvent.Kind.SUBSCRIBE, Bytes.of("HELLO"), null, null, 3);
        }
        throw Errors.unexpected(r, "HELLO 3 map/array reply");
    }

    /** Block for the next frame (acks + deliveries). Close → Closed, timeout → TimedOut. */
    public PubsubEvent recv() {
        Reply r = remote != null ? remote.readOne() : emb.recv(readTimeout);
        return classify(r);
    }

    /** Block for the next Message/Pmessage, skipping ack frames. */
    public PubsubEvent recvMessage() {
        while (true) {
            PubsubEvent ev = recv();
            if (ev.isMessage()) return ev;
        }
    }

    /** Bound a blocking recv (null clears it). A timeout surfaces as TimedOut. */
    public void setReadTimeout(Duration dur) {
        this.readTimeout = dur;
        if (remote != null) remote.setReadTimeout(dur);
    }

    /** A stream of events, terminating on Closed. */
    public Iterator<PubsubEvent> events() {
        return iterate(false);
    }

    /** A stream of messages (ack frames silently skipped), terminating on Closed. */
    public Iterator<PubsubEvent> messages() {
        return iterate(true);
    }

    private Iterator<PubsubEvent> iterate(boolean messagesOnly) {
        return new Iterator<>() {
            private PubsubEvent nextEv;
            private boolean done;

            @Override
            public boolean hasNext() {
                if (done) return false;
                if (nextEv != null) return true;
                try {
                    nextEv = messagesOnly ? recvMessage() : recv();
                    return true;
                } catch (ClosedException e) {
                    done = true;
                    return false;
                }
            }

            @Override
            public PubsubEvent next() {
                if (!hasNext()) throw new NoSuchElementException();
                PubsubEvent ev = nextEv;
                nextEv = null;
                return ev;
            }
        };
    }

    @Override
    public void close() {
        if (remote != null) {
            remote.close();
            remote = null;
        }
        if (emb != null) {
            emb.close();
            emb = null;
        }
    }

    // ---- frame classification (RESP2 array or RESP3 push) ----
    private static PubsubEvent classify(Reply r) {
        if (r.isError()) throw Errors.fromReplyText(r.payload());
        List<Reply> items = r.items();
        if (items == null) throw new ProtocolException("expected array/push frame, got " + r.shape());
        if (items.isEmpty() || items.get(0).payload() == null) {
            throw new ProtocolException("pubsub frame missing kind field");
        }
        String kind = new String(items.get(0).payload(), StandardCharsets.UTF_8);
        return shape(kind, items);
    }

    private static PubsubEvent shape(String kind, List<Reply> items) {
        return switch (kind) {
            case "subscribe" -> ack(PubsubEvent.Kind.SUBSCRIBE, items, false);
            case "psubscribe" -> ack(PubsubEvent.Kind.PSUBSCRIBE, items, true);
            case "unsubscribe" -> ack(PubsubEvent.Kind.UNSUBSCRIBE, items, false);
            case "punsubscribe" -> ack(PubsubEvent.Kind.PUNSUBSCRIBE, items, true);
            case "message" -> {
                if (items.size() != 3) throw new ProtocolException("bad message frame");
                yield new PubsubEvent(PubsubEvent.Kind.MESSAGE, items.get(1).payload(), null, items.get(2).payload(), 0);
            }
            case "pmessage" -> {
                if (items.size() != 4) throw new ProtocolException("bad pmessage frame");
                yield new PubsubEvent(PubsubEvent.Kind.PMESSAGE, items.get(2).payload(), items.get(1).payload(), items.get(3).payload(), 0);
            }
            default -> throw new ProtocolException("unknown pubsub kind '" + kind + "'");
        };
    }

    private static PubsubEvent ack(PubsubEvent.Kind kind, List<Reply> items, boolean pattern) {
        if (items.size() != 3) throw new ProtocolException("expected 3-element pubsub frame");
        byte[] name = items.get(1).isNil() ? null : items.get(1).payload();
        if (!(items.get(2) instanceof Reply.Int cnt)) throw new ProtocolException("bad pubsub count field");
        byte[] channel = pattern ? null : name;
        byte[] pat = pattern ? name : null;
        return new PubsubEvent(kind, channel, pat, null, cnt.value());
    }
}

File: bindings/csharp/Kevy.Client/Subscriber.cs
// Pub/sub consumer side (contract §3.11). A subscribed connection cannot
// send normal commands, so the consumer is a distinct type; publishing is
// done from a normal KevyClient via Publish. Backends by URL: kevy:// /
// redis:// / tcp:// use a dedicated TCP socket; mem://<name> / file:// use
// the in-process bus. Anonymous mem:// is rejected (no other producer — use
// a named bus). recv accepts both RESP2 array and RESP3 push shapes. Both
// faces: Recv/RecvAsync, RecvMessage/RecvMessageAsync.

namespace Kevy;

/// <summary>One subscribed connection — the pub/sub consumer (§3.11).</summary>
public sealed class Subscriber : IDisposable
{
    private RespConnection? _remote;
    private EmbeddedBus? _emb;
    private TimeSpan? _readTimeout;

    private Subscriber() { }

    /// <summary>Open a subscriber without subscribing yet. Anonymous mem://
    /// is rejected.</summary>
    public static Subscriber Connect(string url)
    {
        var t = Url.Parse(url);
        var s = new Subscriber();
        switch (t.Kind)
        {
            case TargetKind.MemAnon:
                throw new KevyUnsupportedException(
                    "anonymous mem:// has no other producer; use mem://<name> for a shared bus");
            case TargetKind.MemNamed or TargetKind.File:
                var (db, key) = Registry.Resolve(t);
                s._emb = new EmbeddedBus(db, key);
                break;
            default:
                s._remote = RespConnection.Dial(t.Host, t.Port);
                break;
        }
        return s;
    }

    /// <summary>Connect and subscribe in one step (≥1 channel required).</summary>
    public static Subscriber ConnectChannels(string url, params KevyBytes[] channels)
    {
        if (channels.Length == 0)
            throw new KevyInvalidInputException(
                "ConnectChannels needs ≥ 1 channel — use Connect for an empty start");
        var s = Connect(url);
        try { s.Subscribe(channels); }
        catch { s.Dispose(); throw; }
        return s;
    }

    /// <summary>SUBSCRIBE to channels (≥1).</summary>
    public void Subscribe(params KevyBytes[] channels)
    {
        if (channels.Length == 0) throw new KevyInvalidInputException("SUBSCRIBE needs ≥ 1 channel");
        if (_remote is not null) _remote.Write(Argv.Cmd("SUBSCRIBE", channels.Raw()));
        else _emb!.Add(channels.Raw(), false);
    }

    /// <summary>PSUBSCRIBE to glob patterns (≥1).</summary>
    public void Psubscribe(params KevyBytes[] patterns)
    {
        if (patterns.Length == 0) throw new KevyInvalidInputException("PSUBSCRIBE needs ≥ 1 pattern");
        if (_remote is not null) _remote.Write(Argv.Cmd("PSUBSCRIBE", patterns.Raw()));
        else _emb!.Add(patterns.Raw(), true);
    }

    /// <summary>UNSUBSCRIBE from channels (empty = all).</summary>
    public void Unsubscribe(params KevyBytes[] channels)
    {
        if (_remote is not null) _remote.Write(Argv.Cmd("UNSUBSCRIBE", channels.Raw()));
        else _emb!.Remove(channels.Raw(), false);
    }

    /// <summary>PUNSUBSCRIBE from patterns (empty = all).</summary>
    public void Punsubscribe(params KevyBytes[] patterns)
    {
        if (_remote is not null) _remote.Write(Argv.Cmd("PUNSUBSCRIBE", patterns.Raw()));
        else _emb!.Remove(patterns.Raw(), true);
    }

    /// <summary>Negotiate RESP3 push frames; remote-only, and must precede
    /// any subscribe (§3.11).</summary>
    public PubsubEvent Hello3()
    {
        var conn = _remote ?? throw new KevyUnsupportedException(
            "HELLO 3 is remote/TCP-only; the embedded bus has no proto switch");
        conn.Write(["HELLO"u8.ToArray(), "3"u8.ToArray()]);
        var r = conn.ReadReply();
        return r.Kind switch
        {
            ReplyKind.Map or ReplyKind.Array => new PubsubEvent(PubsubKind.Subscribe, "HELLO"u8.ToArray(), null, null, 3),
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw new KevyProtocolException("unexpected HELLO 3 reply shape: " + r.Shape()),
        };
    }

    /// <summary>Block for the next frame (acks and deliveries). Close →
    /// Closed; read-timeout → TimedOut.</summary>
    public PubsubEvent Recv() =>
        PubsubFrame.Classify(_remote is not null ? _remote.ReadReply() : Reply.Decode(_emb!.Recv(_readTimeout)));

    /// <summary>Async <see cref="Recv"/>.</summary>
    public async Task<PubsubEvent> RecvAsync(CancellationToken ct = default)
    {
        if (_remote is not null) return PubsubFrame.Classify(await _remote.ReadReplyAsync(ct).ConfigureAwait(false));
        return await Task.Run(() => PubsubFrame.Classify(Reply.Decode(_emb!.Recv(_readTimeout))), ct).ConfigureAwait(false);
    }

    /// <summary>Block for the next Message/Pmessage, skipping ack frames.</summary>
    public (byte[] Channel, byte[] Payload) RecvMessage()
    {
        while (true)
        {
            var ev = Recv();
            if (ev.Kind is PubsubKind.Message or PubsubKind.Pmessage) return (ev.Channel!, ev.Payload!);
        }
    }

    /// <summary>Async <see cref="RecvMessage"/>.</summary>
    public async Task<(byte[] Channel, byte[] Payload)> RecvMessageAsync(CancellationToken ct = default)
    {
        while (true)
        {
            var ev = await RecvAsync(ct).ConfigureAwait(false);
            if (ev.Kind is PubsubKind.Message or PubsubKind.Pmessage) return (ev.Channel!, ev.Payload!);
        }
    }

    /// <summary>Bound a blocking Recv (null clears it). A timeout surfaces
    /// as TimedOut.</summary>
    public void SetReadTimeout(TimeSpan? d)
    {
        _readTimeout = d;
        _remote?.SetReadTimeout(d);
    }

    /// <summary>Stream frames until the connection closes (acks + deliveries).</summary>
    public IEnumerable<PubsubEvent> Events()
    {
        while (true)
        {
            PubsubEvent ev;
            try { ev = Recv(); }
            catch (KevyClosedException) { yield break; }
            yield return ev;
        }
    }

    /// <summary>Stream Message/Pmessage payloads (ack frames skipped).</summary>
    public IEnumerable<(byte[] Channel, byte[] Payload)> Messages()
    {
        foreach (var ev in Events())
            if (ev.Kind is PubsubKind.Message or PubsubKind.Pmessage)
                yield return (ev.Channel!, ev.Payload!);
    }

    /// <summary>Tear down the subscriber.</summary>
    public void Dispose()
    {
        if (_remote is not null) { _remote.Dispose(); _remote = null; }
        if (_emb is not null) { _emb.Dispose(); _emb = null; }
    }
}
